File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    answer = 0
    
    queue = deque()
    queue.append([begin, 0])
    visit = [False for i in range(len(words))]
    
    while queue:
        word, cnt = queue.popleft()
        if word == target:
            return cnt
        for i in range(len(words)):
            tmp = 0
            if not visit[i]:
                for j in range(len(word)):
                    if word[j] != words[i][j]:
                        tmp += 1
                if tmp == 1:
                    queue.append([words[i], cnt+1])
                    visit[i] = True
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;
class Solution {
    static ArrayDeque<Word> q = new ArrayDeque<>();
    static class Word{
        String cur;
        int cnt;
        public Word(String cur, int cnt){
            this.cur = cur;
            this.cnt = cnt;
        }
    }
    public int bfs(String begin, String target, String[] words, boolean[] visited){
        
        q.offer(new Word(begin, 0));
        while(!q.isEmpty()){
            Word word = q.poll();
            if(word.cur.equals(target)){
                return word.cnt;
            }
            
            for(int i = 0; i < words.length; i++){
                int tmp = 0;
                if(visited[i]) continue;
                for(int j = 0; j < words[i].length(); j++){
                    if(word.cur.charAt(j) != words[i].charAt(j)){
                        tmp++;
                    }
                }
                if(tmp == 1){
                    visited[i] = true;
                    q.offer(new Word(words[i], word.cnt + 1));
                }   
            }
        }
        
        return 0;
    }
    public int solution(String begin, String target, String[] words) {
        boolean[] visited = new boolean[words.length];
        return bfs(begin, target, words, visited);
    }
}